extensions/amp-date-countdown/0.1/amp-date-countdown.js

```diff
@@ -0,0 +1,254 @@
+      this.renderItems_(Object.assign(data, localeWordList));
+    }, delay);
+
+    this.registerAction('timeout', this.timeout_.bind(this));
+
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * @param {!Object} items
+   * @return {!Promise}
+   * @private
+   */
+  renderItems_(items) {
+    return this.templates_.findAndRenderTemplate(this.element, items)
+        .then(this.boundRendered_);
+  }
+
+  /**
+   * @return {number}
+   * @private
+   */
+  getEpoch_() {
+    let epoch;
+
+    if (this.endDate_) {
+      epoch = Date.parse(this.endDate_);
+    } else if (this.timestampMs_) {
+      epoch = this.timestampMs_;
+    } else if (this.timestampSeconds_) {
+      epoch = this.timestampSeconds_ * 1000;
+    }
+
+    if (isNaN(epoch)) {
+      user().error(TAG,
+          'One of end-date, timestamp-ms, timestamp-seconds is required');
+    }
+
+    return epoch;
+  }
+
+  /**
+   * Toggles the timeout_ state when timer hit 0.
+   * @param {?../../../src/service/action-impl.ActionInvocation=} opt_invocation
+   * @private
+   */
+  timeout_() {
+
+  }
+
+  /**
+   * @param {string} locale
+   * @return {Object}
+   * @private
+   */
+  getLocaleWord_(locale) {
+    //https://ctrlq.org/code/19899-google-translate-languages refer to google code
+    const localeWord = {
+      'zh-CN': ['年', '月', '天', '小时', '分钟', '秒'],
+      'zh-TW': ['年', '月', '天', '小時', '分鐘', '秒'],
+      'de': ['Jahren', 'Monaten', 'Tagen', 'Stunden', 'Minuten', 'Sekunden'],
+      'en': ['Years', 'Months', 'Days', 'Hours', 'Minutes', 'Seconds'],
+      'es': ['años', 'meses', 'días', 'horas', 'minutos', 'segundos'],
+      'fr': ['ans', 'mois', 'jours', 'heures', 'minutes', 'secondes'],
+      'id': ['tahun', 'bulan', 'hari', 'jam', 'menit', 'detik'],
+      'it': ['anni', 'mesi', 'giorni', 'ore', 'minuti', 'secondi'],
+      'ja': ['年', 'ヶ月', '日', '時間', '分', '秒'],
+      'ko': ['년', '달', '일', '시간', '분', '초'],
+      'nl': ['jaar', 'maanden', 'dagen', 'uur', 'minuten', 'seconden'],
+      'pt': ['anos', 'meses', 'dias', 'horas', 'minutos', 'segundos'],
+      'ru': ['год', 'месяц', 'день', 'час', 'минута', 'секунда'],
+      'th': ['ปี', 'เดือน', 'วัน', 'ชั่วโมง', 'นาที', 'วินาที'],
+      'tr': ['yıl', 'ay', 'gün', 'saat', 'dakika', 'saniye'],
+      'vi': ['năm', 'tháng', 'ngày', 'giờ', 'phút', 'giây'],
+    };
+    if (localeWord[locale]) {
+      const localeWordList = localeWord[locale];
+      return {
+        'years': localeWordList[0],
+        'months': localeWordList[1],
+        'days': localeWordList[2],
+        'hours': localeWordList[3],
+        'minutes': localeWordList[4],
+        'seconds': localeWordList[5],
+      };
+    } else {
+      user().error(TAG, `Invalid locale '${locale}', return empty locale word`);
+      return {};
+    }
+  }
+
+  /**
+   * @param {number} ms
+   * @return {Object}
+   * @private
+   */
+  getYDHMSFromMs_(ms) {
+
+    const priorityList = {
```

Comment:
Use an enum for this. It can be strings since the number value isn't used in any numeric operations. e.g.

```js
/** @enum {string} */
const TIME_UNIT = {
  DAYS: 'days',
  HOURS: 'hours',
  /* ... */
};

Reply:
it actually does have logic, see the comment as below


    const priorityList = {
      days: 1, 
      hours: 2,
      minutes: 3,
      seconds: 4,
    };

    
    const dMs = ms % MILLISECONDS_IN_DAY;
    const h = priorityList[this.biggestUnit_] == 2 
      ? (Math.trunc((ms) / MILLISECONDS_IN_HOUR)) //if biggest-unit is `hours`
      : priorityList[this.biggestUnit_] < 2
        ? (Math.trunc((dMs) / MILLISECONDS_IN_HOUR)) //if biggest-unit is `days`
        : 0; //if biggest-unit is `minutes` or `seconds`

Reply:
Oh I see now. The code would have the same functionality and would be self-documenting if it was like this:

```js
/** @enum {number} */
const TimeUnit = {
  DAYS: 1,
  HOURS: 2,
  MINUTES: 3,
  SECONDS: 4,
};

const h = priorityList[this.biggestUnit_] == TimeUnit.HOURS
      ? (Math.trunc((ms) / MILLISECONDS_IN_HOUR))
      : priorityList[this.biggestUnit_] < TimeUnit.HOURS
        ? (Math.trunc((ms % MILLISECONDS_IN_DAY) / MILLISECONDS_IN_HOUR))
        : 0;
```

When you use the raw number like `priorityList[this.biggestUnit_] == 2` anyone reading the code has to go back and forth to the map to know what the number means, which makes the code more difficult to read.